src/ir_def/freevars.ml

```diff
@@ -138,6 +138,7 @@ and cases cs : f = unions case cs
 and dec d = match d.it with
   | LetD (p, e) -> fd_of_defs (pat p) +++ exp e
   | VarD (i, t, e) -> fd_of_defs (M.singleton i t) +++ exp e
+  | RefD (i, t, e) -> fd_of_defs (M.singleton i t) +++ lexp e
```

Comment:
The idea is that it produces, like VarD, a pointer to a mutable box, but unlike VarD, it is not allocated and filled with the given value, but instead read off an object.

This can only work of the lexp is not an array index, right? So maybe check that in Check_IR. Or maybe cleaner: change the syntax to take an expression and a field name?

Reply:
Maybe instead of introducing RefD, we could get rid of VarD, use just LetD here, but introduce NewE to create new varbox and ObjRefE to read the mutable box off an object? This would also subsume DeclareD if we have a NewUninitializedE or something like that.

In all cases the result would be a pointer to a MutBox (type Mut t) - quite nice.

Reply:
Yes, `RefD` having a `lexp` in it is too liberal. This was just my first push to get something that works.